aip/0194.md

```diff
@@ -0,0 +1,237 @@
+
+> denoting an element of a set which is unchanged in value when multiplied or
+> otherwise operated on by itself.
+
+One interpretation of this definition is that an operation is idempotent if
+repeating the operation results in the same _response_.
+
+This definition allows for retrying `READ` and `UPDATE` requests, since
+repeated runs always result in the same state and response code. However, this
+definition precludes retrying `CREATE` calls, since `ALREADY_EXISTS` may be
+returned unexpectedly (as in the above example). It also precludes retrying
+`DELETE` calls, since `NOT_FOUND` may be returned unexpectedly.
+
+Another intepretation of this definition is that an operation is idempotent if
+repeating the operation results in the same _state_ (or, value).
+
+This second definition allows for retrying all CRUD operations, since the
+interpretation is only concerned with state, and repeating any CRUD operation
+always results in the same state.
+
+This document's retryability guidances follows the latter interpretation,
+which allows for retrying all CRUD operations. The reason for this is that
+there's much more value for users to retry more operations, with little downside.
+
+Consider the downside to retrying the `CREATE` and `DELETE` operations. For
+users it generally just results in an extra code check:
+
+```
+resp, err := client.Create("Foo)
+if err != nil {
+  if status.Convert(err).Code() == codes.ALREADY_EXISTS { // <- extra code check
+    // It already exists.
+    resp = client.Get("Foo")
+  } else {
+    // An actual error.
+    panic(err)
+  }
+}
+_ = resp // TODO: Use resp.
+```
+
+The above code is quite natural and exists in many codebases already: it is the
+easy answer to asynchrony within a system, and a cheaper/more performant way
+of writing code that doesn't know about state (the alternative is to always
+perform a `READ` before a `CREATE`/`DELETE`, which is superfluous).
+
+In comparison to the relatively small downside, there is a large advantage.
+Users will _almost always_ retry `INTERNAL` and `UNAVAILABLE`. By building it
+into a client library instead, users are saved from implementing retry loops,
+deadlines, exponential backoff, and jitter.
+
+Finally, consider that most users _will have to write the if statement either
+way_. After all, a user who implements retry logic themselves will _still_ run
+into the same problem (and solution) described above.
+
+#### Absolute correctness
+
+There is an argument that some user may need to be absolutely sure of the
+state of a system before retrying a non-`GET` request. Such a user is an
+argument in favour of only retrying `GET` requests; never retrying `UPDATE`,
+`CREATE`, or `DELETE`. However, this set of users likely has very niche
+requirements, and is dwarfed by the set of users for whom retrying is
+advantageous.
+
+A simple answer exists to support this set of users: provide an opt-out for
```

Comment:
This sounds more like a design doc than an AIP, to be honest. AIPs are primarily around guidance rather than very detailed justification with multiple "but what about" sections. Maybe this AIP deserves to be an exception to that; Luke can provide more of a steer.

Reply:
Yeah, this section is starting to rathole a bit. This should be part of the _debate around_ the AIP, and can be captured in the PR. That said, we need to have this debate and have it in writing, so keeping this here while we debate it makes sense.

Reply:
I think that anyone reading this document is going to have Strong Opinions and will want to see this reasoning. Is there an appropriate way we can link to it, or something like that?

Reply:
Yeah, we can figure that out. That has been requested in the past on a couple of the others, and we probably do want to have some sort of "here is the debate".

We might borrow a page from Wikipedia here; they have the "talk page" for each Wikipedia article; we might have the "design debate page" for the more contentious questions where we document the decision and why we made it. (e.g. aip.dev/194/debate)

We can figure out how to do this once we get to a consensus on the decision. (It will likely involve me just taking some of @jadeckler's prose and copy-pasting to another file.)